email/email.go

```diff
@@ -0,0 +1,311 @@
+	case "verify":
+		e.subject, err = render("verify_body", verifyHtmlBodyTemplate, data)
+		if err != nil {
+			return err
+		}
+
+		e.htmlBody, err = render("verify_subject", verifySubjectTemplate, data)
+		if err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("unknown template: %s", templateName)
+	}
+
+	return nil
+}
+
+func (e *email) send() error {
+	// See https://medium.com/@dhanushgopinath/sending-html-emails-using-templates-in-golang-9e953ca32f3d
+	// See https://github.com/zemirco/email/blob/master/email.go
+
+	if e.htmlBody == "" {
+		return fmt.Errorf("empty HTML body")
+	}
+
+	if e.subject == "" {
+		return fmt.Errorf("empty HTML body")
+	}
+
+	from, err := mail.ParseAddress(e.from) // validate from address
+	if err != nil {
+		return fmt.Errorf("error parsing address: %v", err)
+	}
+
+	to, err := mail.ParseAddress(e.to) // validate to address
+	if err != nil {
+		return fmt.Errorf("error parsing to address: %v", err)
+	}
+
+	header := textproto.MIMEHeader{}
+	header.Set(textproto.CanonicalMIMEHeaderKey("from"), e.from)
+	header.Set(textproto.CanonicalMIMEHeaderKey("to"), e.to)
+	header.Set(textproto.CanonicalMIMEHeaderKey("reply-to"), e.replyTo)
+	header.Set(textproto.CanonicalMIMEHeaderKey("content-type"), "text/html; charset=UTF-8")
+	header.Set(textproto.CanonicalMIMEHeaderKey("mime-version"), "1.0")
+	header.Set(textproto.CanonicalMIMEHeaderKey("subject"), e.subject)
+
+	var buffer bytes.Buffer
+
+	// write header
+	for key, value := range header {
+		buffer.WriteString(fmt.Sprintf("%s: %s\r\n", key, value[0]))
+	}
+
+	// write body
+	buffer.WriteString(fmt.Sprintf("\r\n%s", e.htmlBody))
+
+	if disableSendEmail {
+		fmt.Printf("DISABLE_SEND_EMAIL=1, email:\n----\n%s\n----\n", buffer.String())
+		return nil
+	} else {
+		addr := fmt.Sprintf("%s:%s", smtpHost, smtpPort)
+		auth := smtp.PlainAuth("", smtpUsername, smtpPassword, smtpHost)
+		return smtp.SendMail(addr, auth, from.Address, []string{to.Address}, buffer.Bytes())
+	}
+}
+
+func render(name string, templateText string, emailTemplateData interface{}) (string, error) {
+
+	t, err := template.New(name).Funcs(funcMap).Parse(templateText)
+
+	if err != nil {
+		return "", err
+	}
+	buf := bytes.NewBuffer(nil)
+	err = t.Execute(buf, emailTemplateData)
+	if err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+var (
+	disableSendEmail bool
+	smtpHost         string
+	smtpPort         string
+	smtpUsername     string
+	smtpPassword     string
+)
+
+// verifyEmail holds the data required to populate the "verify" email templates
+type verifyEmail struct {
+	Email            string
+	VerificationUrl  string
+	RequestIpAddress string
+	RequestTime      time.Time
+	KeyFingerprint   string
+	KeyCreatedDate   time.Time
+}
+
+// funcMap defines template functions that transform variables into strings in the template
+var funcMap = template.FuncMap{
+	"FormatDateTime": func(t time.Time) string {
+		return t.Format("15:04:05 MST on 02 January 2006")
```

Comment:
Perhaps pernickety, but you drop the `0` on `02` in FormatDate below.

Reply:
:+1: